py/picca/fitter2/control_mpi.py

```diff
@@ -0,0 +1,332 @@
+            if cpu_rank == 0 and num_runs < num_cpus:
+                results = results[:num_runs]
+            return results
+
+        result = []
+        # Run 1D grid
+        if dim==1:
+            par = list(self.chi2.dic_chi2scan.keys())[0]
+            grid = self.chi2.dic_chi2scan[par]['grid']
+            num_runs = len(grid)  # Number of points to be run
+
+            if num_runs <= num_cpus:
+                result = mpi_run_1d_set(num_runs, grid, par)
+            else:
+                j = 0
+                for __ in range(num_runs // num_cpus):
+                    result += [mpi_run_1d_set(num_cpus, grid[j:j+num_cpus], par)]
+                    j += num_cpus
+                rest = num_runs % num_cpus
+                if rest != 0:
+                    result += [mpi_run_1d_set(rest, grid[-rest:], par)]
+
+        # Run 2D grid
+        elif dim == 2:
+            par1 = list(self.chi2.dic_chi2scan.keys())[0]
+            par2 = list(self.chi2.dic_chi2scan.keys())[1]
+            values1 = self.chi2.dic_chi2scan[par1]['grid']
+            values2 = self.chi2.dic_chi2scan[par2]['grid']
+            grid = []
+            for v1 in values1:
+                for v2 in values2:
+                    grid += [[v1, v2]]
+            grid = np.asarray(grid)
+            num_runs = len(grid)  # Number of points to be run
+
+            if num_runs <= num_cpus:
+                result = mpi_run_2d_set(num_runs, grid, par1, par2)
+            else:
+                j = 0
+                for __ in range(num_runs // num_cpus):
+                    result += [mpi_run_2d_set(num_cpus, grid[j:j+num_cpus], par1, par2)]
+                    j += num_cpus
+                rest = num_runs % num_cpus
+                if rest != 0:
+                    result += [mpi_run_2d_set(rest, grid[-rest:], par1, par2)]
+
+        # If we are on CPU #0 concatenate results to match the normal output
+        if cpu_rank == 0 and num_runs > num_cpus:
+            temp = result[0]
+            for res in result[1:]:
+                temp = np.r_[temp, res]
+            result = temp
+
+        self.chi2.dic_chi2scan_result = {}
+        self.chi2.dic_chi2scan_result['params'] = np.asarray(np.append(sorted(self.chi2.best_fit.values), ['fval']))
+        self.chi2.dic_chi2scan_result['values'] = np.asarray(result)
+
+        # Set all parameters to where they were before
+        for d in self.chi2.data:
+            store_d_pars_init = store_data_pars[d.name]['init']
+            store_d_par_error = store_data_pars[d.name]['error']
+            store_d_par_fixed = store_data_pars[d.name]['fixed']
+            for name in d.pars_init.keys():
+                d.pars_init[name] = store_d_pars_init[name]
+            for name in d.par_error.keys():
+                d.par_error[name] = store_d_par_error[name]
+            for name in d.par_fixed.keys():
+                d.par_fixed[name] = store_d_par_fixed[name]
+
+    def _mpi_fastMC(self):
```

Comment:
Some of the code below is duplicated from chi2.fastMC(), I wonder whether there could be a way to minimise duplication. If we were to change the behaviour of one of these functions, we would now need to synchronise them by hand...

Reply:
Removing duplication in this case would require a major code reshuffling, as there is no structure that handles the model computation independently. This problem is not localized to these changes either. The original code for fastMC in chi2.py also reuses code from data.py.